File: lib/ExpediaBookings/src/test/java/com/expedia/bookings/unit/Mocker.java
package com.expedia.bookings.unit;

import com.expedia.bookings.services.DateTimeTypeAdapter;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.mobiata.mocke3.FileSystemOpener;

import org.joda.time.DateTime;

import java.io.File;
import java.util.Map;

import static com.mobiata.mocke3.DispatcherUtilitiesKt.loadMockResponseAndReplaceTemplateParams;

public class Mocker {
	public static <T> T object(Class<T> clazz, String mockName, Map<String, String> params) {
		try {
			String root = new File("../mocked/templates").getCanonicalPath();
			FileSystemOpener opener = new FileSystemOpener(root);
			String jsonString = loadMockResponseAndReplaceTemplateParams(mockName, opener, params);
			Gson gson = new GsonBuilder()
					.registerTypeAdapter(DateTime.class, new DateTimeTypeAdapter())
					.create();
			return gson.fromJson(jsonString, clazz);
		}
		catch (Exception e) {
			return null;
		}
	}
}

File: lib/mocked/mocke3/src/main/kotlin/com/mobiata/mocke3/Mocker.kt
package com.mobiata.mocke3

import com.expedia.bookings.services.DateTimeTypeAdapter
import com.google.gson.GsonBuilder
import org.joda.time.DateTime
import java.io.File

fun <T> mockObject(clazz: Class<T>, mockName: String, params: Map<String, String>? = null): T? {
    return try {
        val jsonString = getJsonStringFromMock(mockName, params)
        val gson = GsonBuilder()
                .registerTypeAdapter(DateTime::class.java, DateTimeTypeAdapter())
                .create()
        gson.fromJson(jsonString, clazz)
    } catch (e: Exception) {
        e.printStackTrace()
        null
    }
}

fun getJsonStringFromMock(mockName: String, params: Map<String, String>?): String {
    var file = File("lib/mocked/templates")
    while (!file.exists()) {
        file = File("../" + file.path)
    }
    val opener = FileSystemOpener(file.canonicalPath)
    val jsonString = loadMockResponseAndReplaceTemplateParams(mockName, opener, params)
    return jsonString
}

object Mocker {
    @JvmStatic
    fun <T> loadMock(clazz: Class<T>, mockName: String, params: Map<String, String>): T? {
        return mockObject(clazz, mockName, params)
    }
}
